aten/src/ATen/native/transformers/attention.cpp

```diff
@@ -658,5 +658,67 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> native_decoder_only_multi_head_attent
   return std::make_tuple(std::move(proj), std::move(qkt), std::move(k), std::move(v));
 }
 
+// Computes scaled dot product attention on query, key and value tensors, using
+// an optional attention mask if passed, and applying dropout if a probability
+// greater than 0.0 is specified.
+//
+// Args:
+//     query (Tensor): Query tensor; shape (N, L, E)
+//     key (Tensor): Key tensor; shape (N, S, E)
+//     value (Tensor): Value tensor; shape (N, S, E)
+//     attn_mask (optional Tensor): Attention mask; shape (N, L, S) or (L, S). Currently, only a boolean mask
+//         is supported, where a value of True indicates that the element *should* take part in attention.
+//     dropout_p (float): Dropout probability; if greater than 0.0, dropout is applied
+//     need_attn_weights (bool): If true, the second return value will contain the attention weights used;
+//         otherwise, the second return value is unspecified
+//     is_causal (bool): If true, assumes causal attention masking; for this case, attn_mask should not be set.
+//         TODO: Consider removing this flag before promoting this function to the public API. It's possible
+//         to get specialized support for causal masks (and other types of masking e.g. local attention / block
+//         sparse masks) via tensor subclassing, allowing for a leaner API.
+//
+// Returns a tuple containing:
+//     output (Tensor): Attention output; shape (N, L, E)
+//     attn_weights (Tensor): Attention weighting; shape (N, L, S)
+//
+// Shape legend:
+//     N: Batch size
+//     S: Source sequence length
+//     L: Target sequence length
+//     E: Embedding dimension
+std::tuple<Tensor, Tensor> _scaled_dot_product_attention(
```

Comment:
Adding a comment on code so we can have a thread here. 
- Is there anything we need to do to check Q,K,V inputs as NestedTensors? ie is any mixture of NestedTensor and regular Tensor fine for Q,K,V? 
- Can we arbitrarily pair NestedTensor Q/K/V and attn_mask and/or is_causal=True?

Reply:
Ahh okay +1 on the nit aspect and can be handled when the use case arises. I was more curious than anything but this makes sense.